native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
+     mfcc_buffer. When mfcc_buffer is full, the timestep is copied to batch_buffer.
+     When batch_buffer is full, we do a single step through the acoustic model
+     and accumulate results in StreamingState::accumulated_logits.
+
+     When fininshStream() is called, we decode the accumulated logits and return
+     the corresponding transcription.
+  */
+  StreamingState* setupStream(unsigned int prealloc_frames, unsigned int sample_rate);
+  void feedAudioContent(StreamingState* ctx, const short* buffer, unsigned int buffer_size);
+  char* finishStream(StreamingState* ctx);
+
+private:
+  /**
+   * @brief Perform decoding of the logits, using basic CTC decoder or
+   *        CTC decoder with KenLM enabled
+   *
+   * @param n_frames       Number of timesteps to deal with
+   * @param logits         Matrix of logits, of dimensions:
+   *                       [n_frames][batch_size][num_classes]
+   *
+   * @return String representing the decoded text.
+   *
+   * @note This method will free @p logits.
+   */
+  char* decode(std::vector<float>& logits);
+
+  /**
+   * @brief Do a single inference step in the acoustic model, with:
+   *          input=mfcc
+   *          input_lengths=[n_frames]
+   *
+   * @param mfcc batch input data
+   * @param n_frames number of timesteps in the data
+   *
+   * @param[out] output_logits Should be large enough to fit
+   *                           aNFrames * alphabet_size floats.
+   */
+  void infer(float* mfcc, int n_frames, std::vector<float>& output_logits);
+
+  void processAudioWindow(StreamingState* ctx, short* buf, unsigned int len);
+  void processMfccWindow(StreamingState* ctx, float* buf, unsigned int len);
+  void addZeroMfccWindow(StreamingState* ctx);
+  void processBatch(StreamingState* ctx, float* buf, unsigned int len);
 };
 
+StreamingState*
+Private::setupStream(unsigned int prealloc_frames,
+                     unsigned int /*sample_rate*/)
+{
+  Status status = session->Run({}, {}, {"initialize_state"}, nullptr);
+
+  if (!status.ok()) {
+    std::cerr << "Error running session: " << status << "\n";
+    return nullptr;
+  }
+
+  StreamingState* ctx = new StreamingState;
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+  ctx->accumulated_logits.reserve(prealloc_frames * BATCH_SIZE * num_classes);
+
+  memset(ctx->audio_buffer, 0, sizeof(ctx->audio_buffer));
+  ctx->audio_buffer_len = 0;
+
+  memset(ctx->mfcc_buffer, 0, sizeof(ctx->mfcc_buffer));
+  ctx->mfcc_buffer_len = MFCC_FEATURES*MFCC_CONTEXT; // initial zero entries
+
+  memset(ctx->batch_buffer, 0, sizeof(ctx->batch_buffer));
+  ctx->batch_buffer_len = 0;
+
+  ctx->skip_next_mfcc = false;
+
+  return ctx;
+}
+
+void
+Private::feedAudioContent(StreamingState* ctx,
+                          const short* buffer,
+                          unsigned int buffer_size)
+{
+  while (buffer_size > 0) {
+    while (buffer_size > 0 && ctx->audio_buffer_len < AUDIO_WIN_LEN_SAMPLES) {
```

Comment:
Since we already have `(buffer_size > 0)` above, why repeat it again here?

Reply:
`buffer_size` is decremented inside the inner loop, so `buffer_size > 0` needs to be a condition on the inner loop. The outer loop handles cases where the length of the data passed to feedAudioContent is larger than the available space in the buffer. In that case, we fill the available space in the buffer, process the entire buffer, shift it by one timestep, then continue filling the available space in the buffer.

Reply:
I'll document this behavior in a comment.